[1/2] More optimized & refactored MoEGEMM

sanchitintel opened the pull request:
## Problem

The Existing MoEGEMM implementation is using duplicated headers from cutlass-sycl to use cutlass Group GEMM (except for commenting out a line in a duplicated `xe_array_epilogue` pertaining to `C` matrix, but then that change destroys the generality of groupgemm).

### Performance issues with the MoEGEMM implementation:

1. `B` matrix was being transposed on CPU
2. There were lots of lists being transferred from Host to Device
3. The current implementation doesn't require transporting `num_tokens_per_expert` from GPU to CPU, so this implementation will be useful towards developing a fully fused implementation in the future.

### Integration issues:
1. It's quite messy because some unnecessary cutlass headers had to be duplicated.
2. Further cutlass headers will have to be duplicated, so this approach is not scalable


## Solution

Reference: https://github.com/intel/cutlass-sycl/pull/520 

1. Do not transpose B matrix
2. Only pass the base `A`, `B`, `C`, `D` matrix pointers to the GPU.
3. Prevent D2H & H2D transfers (except for passing a GPU pointer of pointer which points to a base matrix). 
4. Compute tensor offsets in cutlass kernels
5. Instead of copy-pasting cutlass headers, use a separate cutlass branch, and regularly merge main branch commits to it.
6. The interface looks pretty clean now, at least to me.


## Follow-up

1. Improve performance further
2. Consider renaming GroupGEMM in the API to MoEGEMM (perhaps in this PR)?

## Test Plan

UTs
 - [ ] Measure E2E performance
 - [ ] Measure kernel performance (can't be compared to the previous implementation directly since it pre-transposed B on CPU, and then used Rowwise B in the GEMM computation, whereas B is ColumnWise in this PR. 32X32 transpose copy atom has not been added in cutlass-sycl yet, and performance will become better once it's added).


cc @pengzhao-intel

pengzhao-intel commented:
Any performance data of this PR?

sanchitintel commented:
> Any performance data of this PR?

Hi Patric, the cutlass kernel performance in this PR can't be compared directly to the previous implementation because:

1. The main branch transposed B on CPU, but not in this PR, so we save up on that overhead outside the cutlass kernel.
2. The main branch uses `RowMajor B` in the GEMM computation, whereas `B` is ColumnMajor in this PR. The GEMM kernel performance differs for `RowMajor B` & `ColumnMajor `B`, especially since `32x32` transpose copy atom for 16-bit dtypes doesn't currently exist.
I can measure the standalone cutlass kernel performance with `RowMajor` B with https://github.com/intel/cutlass-sycl/pull/520. Then the comparison will be apple-to-apple, because there's no point in comparing performance of `RowMajor B in main branch` vs. `ColumnMajor B in this branch`.

FWIW, the standalone cutlass performance with `RowMajor B using the standalone kernel` was 12.8% better with `num_experts=16, M_per_expert=256, N=16384, K=5120`.

Thanks!

jikunshang commented:
cc @mayuyuace Please take a look for interface.

sanchitintel commented:
>  how do you avoid these D2H and H2D?

1. In the main branch, strides of `A`, `B`, and `D` matrices of each expert were being computed on CPU & were then being passed to GPU. 
2. Besides that, pointer computation of each expert's  `A`, `B`, and `D` matrix was being done on CPU.
3. `num_tokens_per_expert` need not be sent to the CPU.

Now that computation happens on the GPU, with just `A`, `B`, `C`, and `D` pointers being provided.
Technically, this implementation is still doing an H2D copy of 4 `pointers of pointer to tensor`, but I can remove that one as well, if you'd like.

Thanks!

mayuyuace commented:
LGTM

sanchitintel commented:
Hi @LiangLiang-Ma, 

>  Could you please add your optimization in the cutlass-sycl example first?

Which optimization are you referring to? That example is also not using D2H or H2D copies accept for 4 pointers to pointers.
It has the same optimizations as listed in https://github.com/vllm-project/vllm-xpu-kernels/pull/48#issuecomment-3344909093.
The example may have some dead code, though, but it should be okay, as it's not an integration example.


As for this PR, please compare E2E performance or the kernel performance (but with RowMajor B).
This PR's implementation is faster on an Arc B580.

My second PR will improve the performance further.

Liangliang-Ma commented:
> After eliminating D2H & H2D transfers, pointer computation is done on the GPU. But that mostly affects the latency outside the cutlass kernel.

Yes, Like I said D2H/H2D will be handled along with another kernel to eliminate. We would like a overall optimization on fused moe not only in groupedgemm.

Liangliang-Ma commented:
> I had provided you a cutlass example more than 2 weeks ago, and had requested you to modify it for integration.
Even my branch that computes pointers on GPU (https://github.com/intel/cutlass-sycl/pull/520) is older than your vLLM group GEMM PR.

You can check my vllm Grouped Gemm PR, which was started in Aug and when you provide it 2 weeks ago I begun fixing e2e issues. Why you can not just open the commit list and see the time stamp? 

> You had been working on your own without informing us, keeping us in the dark, and that wasted our time as well

Why dont you give me the link of origin grouped gemm example while you just change the frontend? When I saw your code and found what you did was basically same as reusing example, I was shocked.

sanchitintel commented:
> You can check my vllm Grouped Gemm PR, which was started in Aug and when you provide it 2 weeks ago I begun fixing e2e issues. 

https://github.com/intel/cutlass-sycl/pull/520 does pointer computation on GPU

> Why dont you give me the link of origin grouped gemm example while you just change the frontend?

When Patric told me that `num_tokens_per_expert` MUST be on the GPU, I added the changes to the same branch as https://github.com/intel/cutlass-sycl/pull/520, but with `git reset --soft`. 

>  When I saw your code and found what you did was basically same as reusing example, I was shocked

See the header of your group gemm kernel in vLLM main branch, which is actually for a cutlass-sycl example, and explicitly states it has been copy-pasted. I had added that specific header in cutlass.
All the examples in `codeplay-sycl/examples` are copy-pasted to create more examples.
In fact, nvidia cutlass follows the same convention.

Liangliang-Ma commented:
> so I don't see why you're shocked

Because I worked for vLLM and copied example as a base. And you as a kernel developer also copied example.
My major work lands on integration and I dont think reuse example with moving offset calc from some place to another place would waste you too much time.

We need performant kernel not this case of most file change in frontend.

Liangliang-Ma commented:
Just profiled with llama4-scout moe config with 8192 tokens evenly distributed to experts. This PR takes 21.6ms on first gemm and 11.4ms on second gemm. While main's 16.9/8.6ms. 
The profile code is from #https://github.com/Liangliang-Ma/vllm-xpu-kernels/blob/dev/tests/cutlass/profile_moe.py
@sanchitintel

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg      Self XPU    Self XPU %     XPU total  XPU time avg    # of Calls   Total FLOPs
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                           _xpu_C::cutlass_grouped_gemm         5.54%      54.833ms         6.08%      60.162ms      30.081ms      32.862ms        48.35%      32.862ms      16.431ms             2            --
_ZTSN10syclcompat12experimental6detail13KernelFuncto...         0.00%       0.000us         0.00%       0.000us       0.000us      32.834ms        48.31%      32.834ms      16.417ms             2            --
                                   xetla_fused_moe_gemm         0.20%       1.946ms         0.31%       3.041ms       1.521ms      32.812ms        48.28%      32.812ms      16.406ms             2            --
gpu::xetla::MoEGEMM<sycl::_V1::ext::oneapi::bfloat16...         0.00%       0.000us         0.00%       0.000us       0.000us      32.812ms        48.28%      32.812ms      16.406ms             2            --
                               torch_ipex::silu_and_mul         6.23%      61.708ms         6.31%      62.484ms      31.242ms       2.024ms         2.98%       2.024ms       1.012ms             2            --
at::AtenIpexTypeXPU::impl::op_and_mul_functor<c10::B...         0.00%       0.000us         0.00%       0.000us       0.000us       2.024ms         2.98%       2.024ms       1.012ms             2            --
                 Memcpy D2M (DEVICE -> MEMORY(Unknown))         0.00%       0.000us         0.00%       0.000us       0.000us     109.877us         0.16%     109.877us       2.073us            53            --
                              aten::_local_scalar_dense         1.32%      13.088ms         2.02%      20.019ms     488.278us      82.278us         0.12%      82.278us       2.007us            41            --
                                              aten::cat         1.61%      15.893ms         1.90%      18.796ms       2.685ms      48.015us         0.07%      48.015us       6.859us             7            --
                                            aten::copy_         4.39%      43.469ms         4.88%      48.272ms       3.448ms      38.640us         0.06%      38.640us       2.760us            14            --
                 Memcpy M2D (MEMORY(Unknown) -> DEVICE)         0.00%       0.000us         0.00%       0.000us       0.000us      32.704us         0.05%      32.704us       3.634us             9            --
at::native::xpu::CatArrayBatchedCopyKernelFunctor<c1...         0.00%       0.000us         0.00%       0.000us       0.000us      24.894us         0.04%      24.894us       3.556us             7            --
                            Memcpy H2D (HOST -> DEVICE)         0.00%       0.000us         0.00%       0.000us       0.000us      23.121us         0.03%      23.121us       3.303us             7            --
                                              aten::min         1.78%      17.609ms         1.81%      17.968ms      17.968ms      10.312us         0.02%      10.312us      10.312us             1            --
at::native::xpu::ReduceKernel<1, at::native::xpu::Re...         0.00%       0.000us         0.00%       0.000us       0.000us      10.312us         0.02%      10.312us      10.312us             1            --
                                              aten::max         1.77%      17.491ms         1.80%      17.843ms      17.843ms      10.208us         0.02%      10.208us      10.208us             1            --
at::native::xpu::ReduceKernel<1, at::native::xpu::Re...         0.00%       0.000us         0.00%       0.000us       0.000us      10.208us         0.02%      10.208us      10.208us             1            --
                                               aten::ne        17.99%     178.078ms        18.06%     178.787ms      59.596ms       9.791us         0.01%       9.791us       3.264us             3            --
                                            aten::index        23.36%     231.314ms        25.29%     250.346ms     250.346ms       7.083us         0.01%      25.935us      25.935us             1            --
at::native::xpu::IndexKernelFunctor<at::native::xpu:...         0.00%       0.000us         0.00%       0.000us       0.000us       7.083us         0.01%       7.083us       7.083us             1            --
                                      aten::bitwise_and        10.17%     100.647ms        10.23%     101.257ms     101.257ms       6.979us         0.01%       6.979us       6.979us             1            --
at::native::xpu::VectorizedElementwiseKernel<16, at:...         0.00%       0.000us         0.00%       0.000us       0.000us       6.979us         0.01%       6.979us       6.979us             1            --
at::native::xpu::VectorizedElementwiseKernel<8, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       6.875us         0.01%       6.875us       3.438us             2            --
at::native::xpu::VectorizedElementwiseKernel<2, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       6.769us         0.01%       6.769us       2.256us             3            --
                                            copy_if_xpu         0.35%       3.477ms         1.75%      17.302ms      17.302ms       6.666us         0.01%      14.477us      14.477us             1            --
at::native::xpu::UnrolledElementwiseKernel<at::nativ...         0.00%       0.000us         0.00%       0.000us       0.000us       6.458us         0.01%       6.458us       6.458us             1            --
                                              aten::abs         1.96%      19.379ms         4.00%      39.611ms       9.903ms       6.457us         0.01%      12.914us       3.229us             4            --
at::native::xpu::VectorizedElementwiseKernel<8, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       6.457us         0.01%       6.457us       3.229us             2            --
                                              aten::mul        12.31%     121.909ms        12.39%     122.655ms     122.655ms       6.145us         0.01%       6.145us       6.145us             1        36.000
at::native::xpu::VectorizedElementwiseKernel<16, at:...         0.00%       0.000us         0.00%       0.000us       0.000us       6.145us         0.01%       6.145us       6.145us             1            --
                                              aten::div         0.59%       5.851ms         0.69%       6.859ms       6.859ms       5.937us         0.01%       5.937us       5.937us             1            --
at::native::xpu::VectorizedElementwiseKernel<2, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       5.937us         0.01%       5.937us       5.937us             1            --
at::native::xpu::VectorizedElementwiseKernel<8, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       5.624us         0.01%       5.624us       2.812us             2            --
                                     inclusive_scan_xpu         1.20%      11.897ms         1.24%      12.234ms      12.234ms       5.416us         0.01%       5.416us       5.416us             1            --
at::native::xpu::pstl::KSScanKernelFunctor<1, long*,...         0.00%       0.000us         0.00%       0.000us       0.000us       5.416us         0.01%       5.416us       5.416us             1            --
                                             aten::ceil         2.09%      20.685ms         2.12%      20.992ms      20.992ms       5.000us         0.01%       5.000us       5.000us             1            --
at::native::xpu::VectorizedElementwiseKernel<8, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       5.000us         0.01%       5.000us       5.000us             1            --
                                               aten::gt         0.10%       1.025ms         0.12%       1.151ms     575.524us       4.478us         0.01%       4.478us       2.239us             2            --
                                          aten::nonzero         0.04%     375.541us         1.85%      18.281ms      18.281ms       4.375us         0.01%      18.852us      18.852us             1            --
      at::native::xpu::FlattenIdxtoRealIdxKernelFunctor         0.00%       0.000us         0.00%       0.000us       0.000us       4.375us         0.01%       4.375us       4.375us             1            --
at::native::xpu::pstl::PredictKernelFunctor<long, lo...         0.00%       0.000us         0.00%       0.000us       0.000us       4.062us         0.01%       4.062us       4.062us             1            --
                                               aten::eq         0.03%     256.898us         0.04%     391.620us     391.620us       2.708us         0.00%       2.708us       2.708us             1            --
at::native::xpu::pstl::CopyIfKernelFunctor<long, lon...         0.00%       0.000us         0.00%       0.000us       0.000us       2.604us         0.00%       2.604us       2.604us             1            --
                                               aten::lt         0.01%      77.085us         0.01%     122.278us     122.278us       2.291us         0.00%       2.291us       2.291us             1            --
                                    #####ipex_moegemm_1         1.92%      19.007ms         2.20%      21.809ms      21.809ms       0.000us         0.00%      18.902ms      18.902ms             1            --
                             torch_ipex::fused_moe_gemm         0.00%      33.904us         0.31%       3.075ms       1.538ms       0.000us         0.00%      32.812ms      16.406ms             2            --
                                            aten::empty         0.02%     189.241us         0.02%     189.241us       8.228us       0.000us         0.00%       0.000us       0.000us            23            --
                                  urEnqueueKernelLaunch         1.10%      10.924ms         1.10%      10.924ms     321.301us       0.000us         0.00%       0.000us       0.000us            34            --
                                    #####ipex_moegemm_2         1.43%      14.169ms         1.46%      14.443ms      14.443ms       0.000us         0.00%      13.910ms      13.910ms             1            --
                                        aten::transpose         0.00%      38.969us         0.01%      50.401us      16.800us       0.000us         0.00%       0.000us       0.000us             3            --
                                       aten::as_strided         0.01%     117.465us         0.01%     117.465us       1.013us       0.000us         0.00%       0.000us       0.000us           116            --
                                 @@@@@cutlass_moegemm_1         0.10%       1.008ms         5.91%      58.530ms      58.530ms       0.000us         0.00%      21.642ms      21.642ms             1            --
                                               aten::to         0.00%      45.513us         4.89%      48.440ms       3.027ms       0.000us         0.00%      38.640us       2.415us            16            --
                                         aten::_to_copy         0.01%      87.602us         4.89%      48.394ms       3.457ms       0.000us         0.00%      38.640us       2.760us            14            --
                                    aten::empty_strided         0.00%      34.699us         0.00%      34.699us       2.479us       0.000us         0.00%       0.000us       0.000us            14            --
                                     urEnqueueUSMMemcpy         1.77%      17.559ms         1.77%      17.559ms     254.484us       0.000us         0.00%       0.000us       0.000us            69            --
                                       aten::lift_fresh         0.00%       3.791us         0.00%       3.791us       3.791us       0.000us         0.00%       0.000us       0.000us             1            --
                                          aten::detach_         0.00%       3.507us         0.00%       6.465us       6.465us       0.000us         0.00%       0.000us       0.000us             1            --
                                                detach_         0.00%       2.958us         0.00%       2.958us       2.958us       0.000us         0.00%       0.000us       0.000us             1            --
                                       urUSMDeviceAlloc         0.03%     271.095us         0.03%     271.095us      33.887us       0.000us         0.00%       0.000us       0.000us             8            --
                                 @@@@@cutlass_moegemm_2         0.44%       4.362ms        84.10%     832.643ms     832.643ms       0.000us         0.00%      11.484ms      11.484ms             1            --
                                           aten::select         0.02%     213.297us         0.03%     280.960us       3.267us       0.000us         0.00%       0.000us       0.000us            86            --
                                            aten::slice         0.02%     169.948us         0.02%     201.047us       8.377us       0.000us         0.00%       0.000us       0.000us            24            --
                                         urUSMHostAlloc         0.01%     142.862us         0.01%     142.862us     142.862us       0.000us         0.00%       0.000us       0.000us             1            --
                                            aten::stack         0.00%      32.971us         0.05%     449.183us     449.183us       0.000us         0.00%       9.270us       9.270us             1            --
                                             aten::view         0.00%      17.023us         0.00%      17.023us       5.674us       0.000us         0.00%       0.000us       0.000us             3            --
                                          aten::reshape         0.00%      10.771us         0.00%      21.457us      10.728us       0.000us         0.00%       0.000us       0.000us             2            --
                                         aten::isfinite         0.01%     141.338us        32.44%     321.204ms     321.204ms       0.000us         0.00%      17.290us      17.290us             1            --
                                          aten::resize_         0.01%      60.071us         0.01%      60.071us      15.018us       0.000us         0.00%       0.000us       0.000us             4            --
                                          aten::__and__         0.00%      31.858us        10.23%     101.289ms     101.289ms       0.000us         0.00%       6.979us       6.979us             1            --
                                    aten::masked_select         0.01%     116.955us        25.30%     250.476ms     250.476ms       0.000us         0.00%      25.935us      25.935us             1            --
                                             aten::item         0.01%      90.860us         2.03%      20.110ms     490.494us       0.000us         0.00%      82.278us       2.007us            41            --
                                                aten::t         0.00%       5.865us         0.00%      20.422us      20.422us       0.000us         0.00%       0.000us       0.000us             1            --
                                             aten::set_         0.00%      22.405us         0.00%      22.405us      22.405us       0.000us         0.00%       0.000us       0.000us             1            --
                                           aten::unbind         0.01%      66.861us         0.02%     197.674us      98.837us       0.000us         0.00%       0.000us       0.000us             2            --
                                       aten::is_nonzero         0.00%      13.968us         0.22%       2.141ms     535.170us       0.000us         0.00%      10.206us       2.552us             4            --
                                     aten::resolve_conj         0.00%       5.925us         0.00%       5.925us       0.494us       0.000us         0.00%       0.000us       0.000us            12            --
                                      aten::resolve_neg         0.00%       4.806us         0.00%       4.806us       0.401us       0.000us         0.00%       0.000us       0.000us            12            --
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------

```

sanchitintel commented:
@Liangliang-Ma, for the first GEMM, I get 15.04 ms.

sanchitintel commented:
@Liangliang-Ma, for the first GEMM, I get 15.04 ms.

I have commented several times already that an apple-to-apple comparison would be with `RowMajor B` or `ColumnMajor B`for both implementations, and yet you're displaying data for `ColumnMajor B` for this implementation, and `RowMajor B` for the current implementation.

Liangliang-Ma commented:
> I'm consistently getting 11.33 ms for that number when the number of iterations are 200.
e.g. 11..332, 11.334, 11.337

So we can conclude that with your PR we can get 1% to 3% (somewhere btw it) optimization on grouped gemm kernel compared to cutlass example, right? We will discuss about it after our holidays. Thanks!

sanchitintel commented:
For the standalone cutlass kernel, I always see more than 3%, based on your baseline data
As mentioned in the description, E2E performance will see more gains.
Pointer offset computation on CPU is not necessary. It's not happening in some other kernel but python code written by you.
So, I'm not sure why you say that it happens in another kernel.

FWIW, the standalone kernel performance can also be improved even in this PR, since the absolute performance difference between just the cutlass implementations is higher (it was ~12% for these input shapes).

I'll make more changes.

Happy holidays!

sanchitintel commented:
@Liangliang-Ma, 

Just FYI, if you'd pass `static_cast<const ElementC**>((void*)ptr_D.get()` as the value of `ptr_c` argument for the epilogue,  instead of `nullptr` (it's okay to do so, since `C` tiles don't actually get used when `beta` is `0`), then it'd only be ~0.1 TFLOPs slower, or maybe even less.

Then you would be able to use the cutlass-sycl main branch (not referring to the cutlass branch used by this PR), and all the headers you duplicated from cutlass can be removed (provided you're okay with using the default sycl queue), which would make managing the code easier.
